library/src/blas2/rocblas_tpsv.hpp

```diff
@@ -0,0 +1,304 @@
+#include "logging.h"
+#include "rocblas.h"
+#include "utility.h"
+#include <algorithm>
+#include <cstdio>
+#include <tuple>
+
+namespace
+{
+    rocblas_int
+        packed_matrix_index(bool upper, bool trans, rocblas_int n, rocblas_int row, rocblas_int col)
+    {
+        return upper ? (trans ? ((row * (row + 1) / 2) + col) : ((col * (col + 1) / 2) + row))
+                     : (trans ? (((row * (2 * n - row + 1)) / 2) + (col - row))
+                              : (((col * (2 * n - col + 1)) / 2) + (row - col)));
+    }
+
+    // Uses forward substitution to solve Ax = b. Used for a non-transposed lower-triangular matrix
+    // or a transposed upper-triangular matrix.
+    template <bool CONJ, rocblas_int BLK_SIZE, typename T>
+    __device__ void tpsv_forward_substitution_calc(
+        bool diag, bool trans, int n, const T* A, T* x, rocblas_int incx)
+    {
+        __shared__ T xshared[BLK_SIZE];
+        int          tx = threadIdx.x;
+
+        // main loop - iterate forward in BLK_SIZE chunks
+        for(rocblas_int i = 0; i < n; i += BLK_SIZE)
+        {
+            // cache x into shared memory
+            if(tx + i < n)
+                xshared[tx] = x[(tx + i) * incx];
+
+            __syncthreads();
+
+            // iterate through the current block and solve elements
+            for(rocblas_int j = 0; j < BLK_SIZE; j++)
+            {
+                // If the current block covers more than what's left,
+                // we break early.
+                if(j + i >= n)
+                    break;
+
+                // solve element that can be solved
+                if(tx == j && !diag)
+                {
+                    rocblas_int colA   = j + i;
+                    rocblas_int rowA   = j + i;
+                    rocblas_int indexA = packed_matrix_index(trans, trans, n, rowA, colA);
+                    xshared[tx]        = xshared[tx] / (CONJ ? conj(A[indexA]) : A[indexA]);
+                }
+
+                __syncthreads();
+
+                // for rest of block, subtract previous solved part
+                if(tx > j)
+                {
+                    rocblas_int colA   = j + i;
+                    rocblas_int rowA   = tx + i;
+                    rocblas_int indexA = packed_matrix_index(trans, trans, n, rowA, colA);
+
+                    // Ensure row is in range, and subtract
+                    if(rowA < n)
+                        xshared[tx] -= (CONJ ? conj(A[indexA]) : A[indexA]) * xshared[j];
+                }
+            }
+
+            __syncthreads();
+
+            // apply solved diagonal block to the rest of the array
+            // 1. Iterate down rows
+            for(rocblas_int j = BLK_SIZE + i; j < n; j += BLK_SIZE)
+            {
+                if(tx + j >= n)
+                    break;
+
+                // 2. Sum result (across columns) to be subtracted from original value
+                T val = 0;
+                for(rocblas_int p = 0; p < BLK_SIZE; p++)
```

Comment:
SInce BLK_SIZE is known at compile time I think this could be completely unrolled

Reply:
Maybe not important right now though depending on whether we stick with this algorithm in the future when we are optimizing

Reply:
Actually even better I think you can just insert a #pragma unroll above the for loop here (and anywhere else where the number of loop iterations is known at compile time) and it will unroll them for you instead of you doing it manually.

Reply:
The optimizer seems to do the unrolling of reasonable loops automatically from what Yvan has found.   One should experiment.

Reply:
No changes yet, I might take a look for this PR, but it can always be looked at later during the optimization stage.